pkgs/development/libraries/dnnl/default.nix

```diff
@@ -0,0 +1,39 @@
+{ stdenv, lib, fetchFromGitHub, cmake, substituteAll, bash, autoPatchelfHook, mkl ? null }:
+
+stdenv.mkDerivation rec {
+  pname = "dnnl";
+  version = "1.0.2";
+
+  src = fetchFromGitHub {
+    owner = "intel";
+    repo = "mkl-dnn";
+    rev = "v${version}";
+    sha256 = "0lf14a4j511pydyz78wwbl6hifbq86m3dy119iss9cvz8zzrdyvw";
+  };
+  outputs = [ "out" "dev" "doc" ];
+  patches = [ (substituteAll {
+    src = ./bash-to-sh.patch;
+    inherit bash;
+  }) ];
+
+  nativeBuildInputs = [ cmake ];
+
+  buildInputs = lib.optional (mkl != null) mkl;
```

Comment:
Would something like [this](https://github.com/NixOS/nixpkgs/pull/69454/files#diff-4c48ecaa454daa000c372b8b2ca7cfbeR143) be appropriate here?

Reply:
DNNL can be used outside of a Python context, so I'm afraid that won't work in all circumstances (this doesn't depend on numpy).

One can either rely on MKL being allowed in the unfree list (which allows this to work in both mkl and non-mkl situations), as is done here, or it could be enabled by a dedicated flag and require downstream libraries to opt-in.

It would be nice if Nix had a base environment configuration for these numerics, since generally you want consistent libraries across the whole of the stack rather approaching these on a per-library basis.

Reply:
> One can either rely on MKL being allowed in the unfree list

I'M not sure what you mean by this. As far as I know in the current implementation nix would just always try to fulfill the MKL dependency, throwing an error when unfree packages are disabled. One could get around that by explicitly passing `null`, but a proper switch would be nicer.

Reply:
@timokau is right: a quick test demonstrates that this will fail to build if you don't have the unfree packages enabled.  Since `callPackage` will fill in `mkl` from `all-packages.nix`, the default value inside `dnnl/default.nix` has no effect.

An additional issue is that `dnnl`'s CMake build requires a non-default flag to use MKL: https://github.com/intel/mkl-dnn/blob/master/cmake/options.cmake#L157

This PR follows up on what's implemented here and fixes the above two issues:
https://github.com/NixOS/nixpkgs/pull/74893